subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransformOutputs.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.transform;
+
+import com.google.common.collect.ImmutableList;
+
+import java.io.File;
+
+public class DefaultArtifactTransformOutputs implements ArtifactTransformOutputsInternal {
+
+    private final ImmutableList.Builder<File> outputsBuilder = ImmutableList.builder();
+    private final File workspace;
+
+    public DefaultArtifactTransformOutputs(File workspace) {
+        this.workspace = workspace;
+    }
+
+    @Override
+    public void registerOutput(File outputFile) {
+        outputsBuilder.add(outputFile);
```

Comment:
Do we have thread safety concerns here? I guess it's okay to not be thread-safe here, but it should be called out explicitly.

Reply:
This thing is not thread-safe. I don't think we have any thread safe modification methods on the public API at all. For me, it would feel strange to call out something about thread safety here and not somewhere else.

I also wonder how someone would be able to call the method from multiple threads. Spinning up more threads from an artifact transform seems pretty wrong.

Reply:
Using the worker API would be one way to do so. It's not an option today, but hopefully will be soon. But perhaps your transform invokes an existing multi-threaded tool that has a callback to register its outputs or something, then you have the same problem. I guess with the advent of the worker API we'll need to revisit thread safety in many places.

Reply:
That makes sense. I don't think we need to do anything for this PR. But I feel that we need to address this aspect soon.